### DISCUSSION TITLE
msg to tmp?

### DISCUSSION THREAD

**Discussion by flugruger** [nextgenhealthcare/connect#4849] [Q&A] [answered]
Can you help a Mirth noob with a fundamental question? 
**How does message content transfer from msg to tmp (or vice versa)?**
Is the output side populated automatically? Or do I need to manually do something like tmp = msg.copy()?

(I'm on Mirth 3.9.1)

Thanks!

**Comment by rjkroll** [comment]
The tmp object is the internal representation of the outbound template configured in your
transformer settings. 

It will only be available in the transformer when you have an outbound template configured.

The tmp variable is used when you want to convert a message from one format to a completely different format (e.g.
HL7 v2.x to JSON). Or, it can be used to selectively include pieces of the incoming message and map them into the
outbound message.



```
- msg 
- references the incoming message/data 
- var sendApp = msg[‘MSH’][‘MSH.3’][‘MSH.3.1’].toString(



- tmp 
- references the outgoing message/data 
-  tmp[‘MSH’][‘MSH.3’][‘MSH.3.1’] = “EPIC”;
```

**Reply by pacmano1** [reply]
> Or, it can be used to selectively include pieces of the incoming message and map them into the
> outbound message.

I agree but you can just modify msg and not use tmp at all for some use cases.    For example you are simply forwarding the message somewhere else and need to rewrite some component or field.

**Reply by flugruger** [reply]
Thanks, RJ... I appreciate your sharing your knowledge with a noob!

**Comment by tonygermano** [comment, marked as answer]
The short answer to your question is that Mirth does not transfer content between `msg` and `tmp`. That is entirely up to you.

`msg` exists in every transformer. Except for the case when your inbound data type is Raw, where it contains a string value, `msg` is either a plain javascript object or an e4x xml object representing your message after it has been converted to json or xml, if necessary, and then deserialized.

`tmp` only exists if you specify an outbound template for the transformer. It does the same type of conversion on your template as `msg` does on the inbound message. Your outbound template is always used as the starting outbound message that you will manipulate. It is not simply a sample of a possible message like the inbound template. Anything contained in your outbound template will be part of the final message unless you replace or remove it from `tmp`

At the end of your transformer, if an outbound template was specified then `tmp` (with any modifications made to it during the transformer) is used to create the transformed content. Otherwise, `msg` (with any modifications made to it during the transformer) is used to create the transformed content. The transformed content at this point contains either a raw, xml or json string, and is then converted to the outbound data type to create the encoded content.

In the case where you are doing `tmp = msg.copy()` you probably just want to delete your outbound template and make changes to `msg` directly.

**Reply by flugruger** [reply]
Thanks, Tony. I really appreciate this thoughtful response. People like you make the world better.

**Reply by ChristopherSchultz** [reply]
This is a great explanation. Can someone add this to the user's guide?

**Reply by pacmano1** [reply]
@ChristopherSchultz that would be a new feature/idea to create here on GH I think.
